<a href="https://colab.research.google.com/github/fralfaro/MAT281/blob/main/docs/labs/lab_04.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# MAT281 - Laboratorio N°04


**Objetivo**: Aplicar técnicas intermedias y avanzadas de análisis de datos con pandas utilizando un caso real: el Índice de Libertad de Prensa. Este laboratorio incluye operaciones de limpieza, transformación, combinación de datos, y análisis exploratorio usando `merge`, `groupby`, `concat` y otras funciones fundamentales.




**Descripción del Dataset**

El presente conjunto de datos está orientado al análisis del **Índice de Libertad de Prensa**, una métrica internacional que evalúa el nivel de libertad del que gozan periodistas y medios de comunicación en distintos países. Este índice es recopilado anualmente por la organización **Reporteros sin Fronteras**.

La base de datos contempla observaciones por país y año, e incluye tanto el valor del índice como el ranking correspondiente. A menor puntaje en el índice, mayor nivel de libertad de prensa.

**Diccionario de variables**

| Variable     | Clase    | Descripción                                                                          |
| ------------ | -------- | ------------------------------------------------------------------------------------ |
| `codigo_iso` | carácter | Código ISO 3166-1 alfa-3 que representa a cada país.                                 |
| `pais`       | carácter | Nombre oficial del país.                                                             |
| `anio`       | entero   | Año en que se registró la medición del índice.                                       |
| `indice`     | numérico | Valor numérico del Índice de Libertad de Prensa (menor valor indica mayor libertad). |
| `ranking`    | entero   | Posición relativa del país en el ranking mundial de libertad de prensa.              |


**Fuente original y adaptación pedagógica**

* **Fuente original**: [Reporteros sin Fronteras](https://www.rsf-es.org/), recopilado y publicado a través del portal del [Banco Mundial](https://tcdata360.worldbank.org/indicators/h3f86901f?country=BRA&indicator=32416&viz=line_chart&years=2001,2019).
* **Adaptación educativa**: Los archivos han sido modificados intencionalmente para incorporar desafíos técnicos que permiten aplicar los contenidos abordados en clases, tales como limpieza de datos, normalización, detección de duplicados, y combinación de fuentes.


**Descripción de los archivos disponibles**

* **`libertad_prensa_codigo.csv`**: Contiene los pares `codigo_iso` y `pais`. Incluye intencionalmente un código ISO con dos nombres distintos de país para efectos de limpieza y validación de datos.

* **`libertad_prensa_01.csv`**: Contiene registros de los años **anteriores a 2010**. Incluye las variables `PAIS`, `ANIO`, `INDICE`, y `RANKING` con nombres de columna en **mayúsculas**.

* **`libertad_prensa_02.csv`**: Contiene registros de los años **desde 2010 en adelante**. Estructura similar al archivo anterior, con nombres de columna también en **mayúsculas**.





In [ ]:
import numpy as np
import pandas as pd

# lectura de datos
archivos_anio = [
    'https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_01.csv',
    'https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_02.csv'
 ]
df_codigos = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_codigo.csv')



### 1. Consolidación y limpieza de datos

A partir de los archivos disponibles, realice los siguientes pasos:

**a)** Cree un DataFrame llamado `df_anio` que consolide la información proveniente de los archivos **`libertad_prensa_01.csv`** y **`libertad_prensa_02.csv`**, correspondientes a distintas ventanas de tiempo. Recuerde que ambos archivos tienen nombres de columnas en mayúscula, por lo que debe normalizarlas a **minúscula** para asegurar consistencia.

**b)** Explore el archivo **`libertad_prensa_codigo.csv`** e identifique el código ISO que aparece asociado a dos nombres de país distintos. Elimine el registro que corresponda a un valor incorrecto o inconsistente, conservando solo el que considere válido.

**c)** Una vez preparados los archivos, cree un nuevo DataFrame llamado `df` que combine `df_anio` con `df_codigos`, utilizando la columna `codigo_iso` como clave. Asegúrese de realizar una unión que conserve únicamente los registros que tengan coincidencia en ambas fuentes.

> **Sugerencia**:
>
> * Para unir los archivos por filas (años), utilice la función `pd.concat([...])`.
> * Para combinar información por columnas (variables), utilice `pd.merge(...)` especificando `on='codigo_iso'`.



In [27]:
df_01 = pd.read_csv(archivos_anio[0])
df_02 = pd.read_csv(archivos_anio[1])
df_anio = pd.concat([df_01, df_02], ignore_index=True)
df_anio.columns = df_anio.columns.str.lower()

df_codigos_cleaned = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_codigo.csv')
df_codigos_cleaned = df_codigos_cleaned[~((df_codigos_cleaned['codigo_iso'] == 'ZWE') & (df_codigos_cleaned['pais'] == 'malo'))]

df = pd.merge(df_anio, df_codigos_cleaned, on='codigo_iso', how='inner')
print("DataFrame 'df' consolidado y limpio creado.")


DataFrame 'df' consolidado y limpio creado.




### 2. Exploración inicial del conjunto de datos

Una vez que hayas consolidado el DataFrame final `df`, realiza un análisis exploratorio básico respondiendo las siguientes preguntas:

#### **Estructura del DataFrame**

* ¿Cuántas **filas (observaciones)** contiene el conjunto de datos?
* ¿Cuántas **columnas** tiene el DataFrame?
* ¿Cuáles son los **nombres de las columnas**?
* ¿Qué **tipo de datos** tiene cada columna?
* ¿Hay columnas con un tipo de dato inesperado (por ejemplo, fechas como strings)?

#### **Resumen estadístico**

* Genera un resumen estadístico del conjunto de datos con `.describe()`.
  ¿Qué observas sobre los valores de `indice` y `ranking`?
* ¿Qué valores mínimo, máximo y promedio tiene la columna `indice`?
* ¿Qué países presentan los valores extremos en `indice` y `ranking`?

#### **Datos faltantes**

* ¿Cuántos valores nulos hay en cada columna?
* ¿Qué proporción de observaciones tienen valores faltantes?
* ¿Hay columnas con más del 30% de datos faltantes?

#### **Unicidad y duplicados**

* ¿Cuántos países distintos (`pais`) hay en el DataFrame?
* ¿Cuántos años distintos (`anio`) hay representados?
* ¿Existen filas duplicadas (exactamente iguales)? ¿Cuántas?

#### **Validación cruzada de columnas**

* ¿Hay inconsistencias entre el país (`pais`) y su código (`codigo_iso`)?
  (por ejemplo, un mismo código ISO asociado a más de un país)

> **Sugerencia**: Apoya tu análisis con funciones como `.info()`, `.nunique()`, `.isnull().sum()`, `.duplicated()`, `.value_counts()`, entre otras.



    

In [28]:
print("Información concisa del DataFrame df:")
df.info()
print("\nNúmero de filas y columnas del DataFrame df:")
print(df.shape)
print("\nNombres de las columnas del DataFrame df:")
print(df.columns.tolist())

print("\nResumen estadístico del DataFrame df:")
display(df.describe())

print("\nAnálisis de la columna 'indice':")
print(f"Valor mínimo de indice: {df['indice'].min()}")
print(f"Valor máximo de indice: {df['indice'].max()}")
print(f"Valor promedio de indice: {df['indice'].mean()}")

print("\nAnálisis de la columna 'ranking':")
print(f"Valor mínimo de ranking: {df['ranking'].min()}")
print(f"Valor máximo de ranking: {df['ranking'].max()}")
print(f"Valor promedio de ranking: {df['ranking'].mean()}")

print("\nPaíses con valores extremos en 'indice':")
min_indice_country = df.loc[df['indice'].idxmin(), 'pais']
max_indice_country = df.loc[df['indice'].idxmax(), 'pais']
print(f"País con el indice más bajo: {min_indice_country} (indice = {df['indice'].min()})")
print(f"País con el indice más alto: {max_indice_country} (indice = {df['indice'].max()})")

print("\nPaíses con valores extremos en 'ranking':")
min_ranking_country = df.loc[df['ranking'].idxmin(), 'pais']
max_ranking_country = df.loc[df['ranking'].idxmax(), 'pais']
print(f"País con el ranking más bajo: {min_ranking_country} (ranking = {df['ranking'].min()})")
print(f"País con el ranking más alto: {max_ranking_country} (ranking = {df['ranking'].max()})")

print("\n--- Análisis de Datos Faltantes ---")
nulos_por_columna = df.isnull().sum()
proporcion_nulos = (df.isnull().sum() / len(df)) * 100

print("\nValores nulos por columna:")
print(nulos_por_columna)

print("\nProporción de valores nulos por columna (%):")
print(proporcion_nulos)

columnas_con_muchos_nulos = proporcion_nulos[proporcion_nulos > 30]
if not columnas_con_muchos_nulos.empty:
    print("\nColumnas con más del 30% de datos faltantes:")
    print(columnas_con_muchos_nulos)
else:
    print("\nNo hay columnas con más del 30% de datos faltantes.")

print("\n--- Unicidad y Duplicados ---")
print(f"Número de países distintos: {df['pais'].nunique()}")
print(f"Número de años distintos: {df['anio'].nunique()}")
num_duplicated_rows = df.duplicated().sum()
print(f"Número de filas duplicadas (exactamente iguales): {num_duplicated_rows}")

print("\n--- Validación cruzada de columnas (País-Código ISO) ---")
inconsistent_codes = df.groupby('codigo_iso')['pais'].nunique()
inconsistent_codes = inconsistent_codes[inconsistent_codes > 1]
print("Códigos ISO con inconsistencias (asociados a más de un país):")
print(inconsistent_codes)


Información concisa del DataFrame df:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3060 entries, 0 to 3059
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   codigo_iso  3060 non-null   object 
 1   anio        3060 non-null   int64  
 2   indice      2664 non-null   float64
 3   ranking     2837 non-null   float64
 4   pais        3060 non-null   object 
dtypes: float64(2), int64(1), object(2)
memory usage: 119.7+ KB

Número de filas y columnas del DataFrame df:
(3060, 5)

Nombres de las columnas del DataFrame df:
['codigo_iso', 'anio', 'indice', 'ranking', 'pais']

Resumen estadístico del DataFrame df:


,anio,indice,ranking
count,3060.000000,2664.000000,2837.000000
mean,2009.941176,205.782316,477.930913
std,5.786024,2695.525264,6474.935347
min,2001.000000,0.000000,1.000000
25%,2005.000000,15.295000,34.000000
50%,2009.000000,28.000000,70.000000
75%,2015.000000,41.227500,110.000000
max,2019.000000,64536.000000,121056.000000



Análisis de la columna 'indice':
Valor mínimo de indice: 0.0
Valor máximo de indice: 64536.0
Valor promedio de indice: 205.7823160660661

Análisis de la columna 'ranking':
Valor mínimo de ranking: 1.0
Valor máximo de ranking: 121056.0
Valor promedio de ranking: 477.93091293620023

Países con valores extremos en 'indice':
País con el indice más bajo: Dinamarca (indice = 0.0)
País con el indice más alto: Kosovo (indice = 64536.0)

Países con valores extremos en 'ranking':
País con el ranking más bajo: Finlandia (ranking = 1.0)
País con el ranking más alto: Kosovo (ranking = 121056.0)

--- Análisis de Datos Faltantes ---

Valores nulos por columna:
codigo_iso      0
anio            0
indice        396
ranking       223
pais            0
dtype: int64

Proporción de valores nulos por columna (%):
codigo_iso     0.000000
anio           0.000000
indice        12.941176
ranking        7.287582
pais           0.000000
dtype: float64

No hay columnas con más del 30% de datos faltantes.

--- Uni




### 3. Comparación regional: países latinoamericanos

En esta sección se busca identificar cuáles son los países de América Latina que han presentado los valores extremos del **Índice de Libertad de Prensa** en cada año observado.

> Recuerda que un menor puntaje en `indice` implica mayor libertad de prensa.

#### **Tareas:**

**a)** Utilizando un ciclo `for`, recorre cada año del conjunto de datos filtrado por países latinoamericanos, y determina para cada año:

* El país con el menor valor de `indice` (mayor libertad de prensa).
* El país con el mayor valor de `indice` (menor libertad de prensa).

**b)** Resuelve la misma tarea del punto anterior utilizando un enfoque vectorizado con `groupby`, sin usar ciclos explícitos.



#### **Lista de países latinoamericanos considerada:**

```python
america = ['ARG', 'ATG', 'BLZ', 'BOL', 'BRA', 'CAN', 'CHL', 'COL', 'CRI',
           'CUB', 'DOM', 'ECU', 'GRD', 'GTM', 'GUY', 'HND', 'HTI', 'JAM',
           'MEX', 'NIC', 'PAN', 'PER', 'PRY', 'SLV', 'SUR', 'TTO', 'URY',
           'USA', 'VEN']
```

> Puedes usar esta lista para filtrar el DataFrame final por la columna `codigo_iso`.



In [29]:
america = ['ARG', 'ATG', 'BLZ', 'BOL', 'BRA', 'CAN', 'CHL', 'COL', 'CRI',
           'CUB', 'DOM', 'ECU', 'GRD', 'GTM', 'GUY', 'HND', 'HTI', 'JAM',
           'MEX', 'NIC', 'PAN', 'PER', 'PRY', 'SLV', 'SUR', 'TTO', 'URY',
           'USA', 'VEN']

df_america = df[df['codigo_iso'].isin(america)].copy()
print("DataFrame filtrado para países latinoamericanos (df_america) creado.")
print(f"Shape de df_america: {df_america.shape}")
display(df_america.head())

print("\n--- Comparación Regional (por año) usando un bucle for: ---")
años_unicos = df_america['anio'].sort_values().unique()
for anio in años_unicos:
    df_anio_america = df_america[df_america['anio'] == anio].dropna(subset=['indice'])

    if not df_anio_america.empty:
        min_indice_pais = df_anio_america.loc[df_anio_america['indice'].idxmin()]
        max_indice_pais = df_anio_america.loc[df_anio_america['indice'].idxmax()]

        print(f"\nAño {anio}:")
        print(f"  Mayor libertad de prensa: {min_indice_pais['pais']} (indice: {min_indice_pais['indice']})")
        print(f"  Menor libertad de prensa: {max_indice_pais['pais']} (indice: {max_indice_pais['indice']})")
    else:
        print(f"\nAño {anio}: No hay datos de 'indice' disponibles para este año en la región.")

print("\n--- Comparación Regional (por año) usando un enfoque vectorizado: ---")
df_america_filtered = df_america.dropna(subset=['indice'])

idx_min = df_america_filtered.groupby('anio')['indice'].idxmin()
paises_mayor_libertad = df_america_filtered.loc[idx_min]

idx_max = df_america_filtered.groupby('anio')['indice'].idxmax()
paises_menor_libertad = df_america_filtered.loc[idx_max]

for anio in paises_mayor_libertad['anio'].unique():
    mayor_libertad_pais = paises_mayor_libertad[paises_mayor_libertad['anio'] == anio].iloc[0]
    menor_libertad_pais = paises_menor_libertad[paises_menor_libertad['anio'] == anio].iloc[0]

    print(f"\nAño {anio}:")
    print(f"  Mayor libertad de prensa: {mayor_libertad_pais['pais']} (indice: {mayor_libertad_pais['indice']})")
    print(f"  Menor libertad de prensa: {menor_libertad_pais['pais']} (indice: {menor_libertad_pais['indice']})")

años_con_datos = df_america_filtered['anio'].unique()
años_unicos_total = df_america['anio'].sort_values().unique()

for anio in años_unicos_total:
    if anio not in años_con_datos:
        print(f"\nAño {anio}: No hay datos de 'indice' disponibles para este año en la región.")


DataFrame filtrado para países latinoamericanos (df_america) creado.
Shape de df_america: (493, 5)


,codigo_iso,anio,indice,ranking,pais
5,ARG,2001,12.0,8.0,Argentina
7,ATG,2001,NaN,NaN,Antigua y Barbuda
20,BLZ,2001,NaN,NaN,Belize
21,BOL,2001,14.5,13.0,Bolivia
22,BRA,2001,18.8,18.0,Brasil



--- Comparación Regional (por año) usando un bucle for: ---

Año 2001:
  Mayor libertad de prensa: Canadá (indice: 0.8)
  Menor libertad de prensa: Cuba (indice: 90.3)

Año 2002:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 1.0)
  Menor libertad de prensa: Cuba (indice: 97.83)

Año 2003:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 2.0)
  Menor libertad de prensa: Argentina (indice: 35826.0)

Año 2004:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 2.0)
  Menor libertad de prensa: Cuba (indice: 87.0)

Año 2005:
  Mayor libertad de prensa: Bolivia (indice: 4.5)
  Menor libertad de prensa: Cuba (indice: 95.0)

Año 2006:
  Mayor libertad de prensa: Canadá (indice: 4.88)
  Menor libertad de prensa: Cuba (indice: 96.17)

Año 2007:
  Mayor libertad de prensa: Canadá (indice: 3.33)
  Menor libertad de prensa: Cuba (indice: 88.33)

Año 2008:
  Mayor libertad de prensa: Canadá (indice: 3.7)
  Menor libertad de prensa: Cuba (indice: 94.0)

Año 2009:
  Mayor libertad 

### 4. Análisis anual del índice por país

En esta sección se busca analizar la evolución del **índice máximo** de libertad de prensa alcanzado por cada país a lo largo del tiempo.

#### **Tarea principal:**

* Construye una tabla dinámica (`pivot_table`) donde las **filas** correspondan a los países, las **columnas** a los años (`anio`) y los **valores** sean el `indice` máximo alcanzado por cada país en ese año.
* Asegúrate de reemplazar los valores nulos resultantes con `0`.

> **Hint**: Puedes utilizar el parámetro `fill_value=0` en `pd.pivot_table(...)`.



#### **Preguntas adicionales:**

**a)** ¿Qué país tiene el mayor valor de `indice` en toda la tabla resultante? ¿Y cuál tiene el menor (distinto de cero)?
**b)** ¿Qué años presentan en promedio los valores de `indice` más altos? ¿Y los más bajos?

> (Pista: usa `.mean(axis=0)` sobre la tabla pivot)

**c)** ¿Qué país muestra mayor **variabilidad** (diferencia entre su máximo y mínimo `indice` a lo largo del tiempo)?

> (Pista: aplica `.max(axis=1) - .min(axis=1)`)

**d)** ¿Existen países con índice constante a lo largo de todos los años registrados? ¿Cuáles?

**e)** ¿Qué países no tienen ningún dato (es decir, quedaron con todos los valores igual a 0)? ¿Podrías explicar por qué?





### Respuesta a la pregunta a):

*   **País con el mayor `indice`**: **Kosovo** en el año 2014, con un `indice` de 64536.0.
*   **País con el menor `indice` (distinto de cero)**: **Austria** en el año 2009, con un `indice` de 0.5.

### Respuesta a la pregunta b):

*   **Año con el promedio de `indice` más alto**: **2013** (Promedio: 449.11).
*   **Año con el promedio de `indice` más bajo**: **2001** (Promedio: 20.03).

### Respuesta a la pregunta c):

*   El país que muestra mayor **variabilidad** (diferencia entre su máximo y mínimo `indice` a lo largo del tiempo) es **Kosovo** (Variabilidad: 46098.00).

### Respuesta a la pregunta d):

*   Sí, existen países con un `indice` constante a lo largo de los años registrados con datos. Estos son: **Antigua y Barbuda** y **Granada**.

### Respuesta a la pregunta e):

*   No se encontraron países sin ningún dato (es decir, con todos los valores en 0) en la tabla dinámica. **Todos los países tienen al menos un dato.** Los ceros en la tabla dinámica representan años para los cuales el país no tenía un `indice` registrado en los datos originales, y esos `NaN`s fueron reemplazados por 0 al construir la tabla dinámica.

### Países con índice constante

Sí, se identificaron **Antigua y Barbuda** y **Granada** como los países con un `indice` constante (sin variabilidad) a lo largo de los años registrados con datos.

### Países sin datos (todos los valores en cero)

No se encontraron países que no tuvieran ningún dato (es decir, con todos los valores en 0) en la tabla dinámica. Todos los países tienen al menos un dato registrado.

In [30]:
import numpy as np

print("--- Tabla dinámica (pivot_df) de índice anual por país ---")
pivot_df = pd.pivot_table(df, values='indice', index='pais', columns='anio', aggfunc='max', fill_value=0)
display(pivot_df.head())

print("\n--- Análisis de Tabla Dinámica (Extremos) ---")
max_indice_global = pivot_df.max().max()
min_indice_global = pivot_df[pivot_df > 0].min().min() # Excluir ceros

max_indice_pais_anio = pivot_df.stack().idxmax()
max_pais = max_indice_pais_anio[0]
max_anio = max_indice_pais_anio[1]

min_indice_pais_anio = pivot_df[pivot_df > 0].stack().idxmin()
min_pais = min_indice_pais_anio[0]
min_anio = min_indice_pais_anio[1]

print(f"País con el mayor 'indice' en toda la tabla: {max_pais} en el año {max_anio} con un indice de {max_indice_global}")
print(f"País con el menor 'indice' (distinto de cero) en toda la tabla: {min_pais} en el año {min_anio} con un indice de {min_indice_global}")

print("\n--- Análisis de Tabla Dinámica (Años) ---")
promedio_indice_por_anio = pivot_df.mean(axis=0)

anio_mas_alto_promedio = promedio_indice_por_anio.idxmax()
valor_mas_alto_promedio = promedio_indice_por_anio.max()

anio_mas_bajo_promedio = promedio_indice_por_anio.idxmin()
valor_mas_bajo_promedio = promedio_indice_por_anio.min()

print(f"Año con el promedio de 'indice' más alto: {anio_mas_alto_promedio} (Promedio: {valor_mas_alto_promedio:.2f})")
print(f"Año con el promedio de 'indice' más bajo: {anio_mas_bajo_promedio} (Promedio: {valor_mas_bajo_promedio:.2f})")

print("\n--- Análisis de Tabla Dinámica (Variabilidad) ---")
max_indice_por_pais = pivot_df.replace(0, np.nan).max(axis=1)
min_indice_por_pais = pivot_df.replace(0, np.nan).min(axis=1)

variabilidad_por_pais = max_indice_por_pais - min_indice_por_pais

pais_mayor_variabilidad = variabilidad_por_pais.idxmax()
valor_mayor_variabilidad = variabilidad_por_pais.max()

print(f"País con mayor variabilidad en 'indice': {pais_mayor_variabilidad} (Variabilidad: {valor_mayor_variabilidad:.2f})")

print("\n--- Análisis de Tabla Dinámica (Constancia y Nulos) ---")
paises_indice_constante = variabilidad_por_pais[variabilidad_por_pais == 0].index.tolist()

print("Países con índice constante (sin variabilidad a lo largo de los años con datos):")
if paises_indice_constante:
    for pais in paises_indice_constante:
        print(f"- {pais}")
else:
    print("Ningún país tiene un índice constante.")

paises_sin_datos = pivot_df[pivot_df.sum(axis=1) == 0].index.tolist()

print("\nPaíses sin ningún dato (todos los valores en 0):")
if paises_sin_datos:
    for pais in paises_sin_datos:
        print(f"- {pais}")
else:
    print("Todos los países tienen al menos un dato.")


--- Tabla dinámica (pivot_df) de índice anual por país ---


anio,2001,2002,2003,2004,2005,2006,2007,2008,2009,2012,2013,2014,2015,2017,2018,2019
pais,,,,,,,,,,,,,,,,
Afghanistán,35.5,40.17,28.25,39.17,44.25,56.50,59.25,54.25,51.67,37.36,37.07,37.44,37.75,39.46,37.28,36.55
Albania,0.0,6.50,11.50,14.17,18.00,25.50,16.00,21.75,21.50,30.88,29.92,28.77,29.92,29.92,29.49,29.84
Alemania,1.5,1.33,2.00,4.00,5.50,5.75,4.50,3.50,4.25,10.24,10.23,11.47,14.80,14.97,14.39,14.60
Algeria,31.0,33.00,43.50,40.33,40.00,40.50,31.33,49.56,47.33,36.54,36.26,36.63,41.69,42.83,43.13,45.75
Andorra,0.0,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,6.82,6.82,19.87,19.87,21.03,22.21,24.63



--- Análisis de Tabla Dinámica (Extremos) ---
País con el mayor 'indice' en toda la tabla: Kosovo en el año 2014 con un indice de 64536.0
País con el menor 'indice' (distinto de cero) en toda la tabla: Austria en el año 2009 con un indice de 0.5

--- Análisis de Tabla Dinámica (Años) ---
Año con el promedio de 'indice' más alto: 2013 (Promedio: 449.11)
Año con el promedio de 'indice' más bajo: 2001 (Promedio: 20.03)

--- Análisis de Tabla Dinámica (Variabilidad) ---
País con mayor variabilidad en 'indice': Kosovo (Variabilidad: 46098.00)

--- Análisis de Tabla Dinámica (Constancia y Nulos) ---
Países con índice constante (sin variabilidad a lo largo de los años con datos):
- Antigua y Barbuda
- Granada

Países sin ningún dato (todos los valores en 0):
Todos los países tienen al menos un dato.


# Task
The user wants to analyze the Press Freedom Index data. The task involves consolidating and cleaning data from multiple CSV files, performing an initial exploratory data analysis, comparing regional data for Latin American countries, and finally analyzing the annual index by country using a pivot table. The analysis includes identifying extreme values, variabilities, and constant indices.

## Consolidar, Limpiar y Combinar Datos

### Subtask:
Consolidate data from 'libertad_prensa_01.csv' and 'libertad_prensa_02.csv' into `df_anio`, normalize column names to lowercase, clean 'libertad_prensa_codigo.csv' by removing the inconsistent country name, and then merge `df_anio` with the cleaned codes DataFrame into a final DataFrame `df`.


**Reasoning**:
I will load the two CSV files into DataFrames, concatenate them, and then convert the column names to lowercase to create `df_anio`.



In [ ]:
df_01 = pd.read_csv(archivos_anio[0])
df_02 = pd.read_csv(archivos_anio[1])

df_anio = pd.concat([df_01, df_02], ignore_index=True)
df_anio.columns = df_anio.columns.str.lower()

print("df_anio created successfully with consolidated data and lowercase column names.")
df_anio.head()

**Reasoning**:
To identify the inconsistent `codigo_iso` and `pais` combinations, I will display the value counts for the `codigo_iso` column in `df_codigos`.



In [1]:
df_codigos['codigo_iso'].value_counts()

NameError: name 'df_codigos' is not defined

**Reasoning**:
I need to ensure `df_codigos` is defined. I will re-load `df_codigos` from its source and then display the value counts for the `codigo_iso` column to identify inconsistencies.



In [3]:
import pandas as pd
df_codigos = pd.read_csv('https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_codigo.csv')
df_codigos['codigo_iso'].value_counts()

,count
codigo_iso,
ZWE,2
AFG,1
ALB,1
AGO,1
ARE,1
...,...
VEN,1
WSM,1
YEM,1


**Reasoning**:
To identify which country name associated with 'ZWE' is incorrect, I will filter the `df_codigos` DataFrame to show only the rows where `codigo_iso` is 'ZWE'.



In [4]:
df_codigos[df_codigos['codigo_iso'] == 'ZWE']

,codigo_iso,pais
179,ZWE,Zimbabue
180,ZWE,malo


**Reasoning**:
I need to remove the incorrect entry for 'ZWE' in `df_codigos`. I will filter out the row where `codigo_iso` is 'ZWE' and `pais` is 'malo'.



In [5]:
df_codigos = df_codigos[~((df_codigos['codigo_iso'] == 'ZWE') & (df_codigos['pais'] == 'malo'))]
print("Incorrect entry for 'ZWE' removed from df_codigos.")
df_codigos.head()

Incorrect entry for 'ZWE' removed from df_codigos.


,codigo_iso,pais
0,AFG,Afghanistán
1,AGO,Angola
2,ALB,Albania
3,AND,Andorra
4,ARE,Emiratos Árabes Unidos


**Reasoning**:
Now that both `df_anio` and `df_codigos` have been prepared, I will merge them into a new DataFrame `df` using `codigo_iso` as the key and performing an inner join, as specified in the subtask instructions. This will create the final consolidated and cleaned DataFrame.



In [8]:
archivos_anio = [
    'https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_01.csv',
    'https://raw.githubusercontent.com/fralfaro/MAT281/main/docs/labs/data/libertad_prensa_02.csv'
 ]
df_01 = pd.read_csv(archivos_anio[0])
df_02 = pd.read_csv(archivos_anio[1])

df_anio = pd.concat([df_01, df_02], ignore_index=True)
df_anio.columns = df_anio.columns.str.lower()

print("df_anio created successfully with consolidated data and lowercase column names.")

df = pd.merge(df_anio, df_codigos, on='codigo_iso', how='inner')
print("df created successfully by merging df_anio and df_codigos.")
df.head()

df_anio created successfully with consolidated data and lowercase column names.
df created successfully by merging df_anio and df_codigos.


,codigo_iso,anio,indice,ranking,pais
0,AFG,2001,35.5,59.0,Afghanistán
1,AGO,2001,30.2,50.0,Angola
2,ALB,2001,NaN,NaN,Albania
3,AND,2001,NaN,NaN,Andorra
4,ARE,2001,NaN,NaN,Emiratos Árabes Unidos


## Explorar Estructura del DataFrame

### Subtask:
Analizar la estructura del DataFrame `df` para determinar el número de filas, columnas, nombres de columnas, tipos de datos y verificar inconsistencias.


**Reasoning**:
Displaying the DataFrame's info, shape, and columns as requested by the instructions to explore its structure.



In [9]:
print("Información concisa del DataFrame df:")
df.info()
print("\nNúmero de filas y columnas del DataFrame df:")
print(df.shape)
print("\nNombres de las columnas del DataFrame df:")
print(df.columns.tolist())

Información concisa del DataFrame df:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3060 entries, 0 to 3059
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   codigo_iso  3060 non-null   object 
 1   anio        3060 non-null   int64  
 2   indice      2664 non-null   float64
 3   ranking     2837 non-null   float64
 4   pais        3060 non-null   object 
dtypes: float64(2), int64(1), object(2)
memory usage: 119.7+ KB

Número de filas y columnas del DataFrame df:
(3060, 5)

Nombres de las columnas del DataFrame df:
['codigo_iso', 'anio', 'indice', 'ranking', 'pais']


# Task
Analyze the Press Freedom Index data by consolidating and cleaning data from multiple CSV files, performing an initial exploratory data analysis, comparing regional data for Latin American countries, and finally analyzing the annual index by country using a pivot table. The analysis should include identifying extreme values, variabilities, and constant indices.

## Generar Resumen Estadístico

### Subtask:
Obtener un resumen estadístico del DataFrame `df`, analizar los valores mínimo, máximo y promedio de las columnas `indice` y `ranking`, y identificar países con valores extremos en estas columnas.


**Reasoning**:
Generate a statistical summary of the DataFrame `df` using the `.describe()` method to observe the values of `indice` and `ranking`.



In [10]:
print("Resumen estadístico del DataFrame df:")
df.describe()

Resumen estadístico del DataFrame df:


,anio,indice,ranking
count,3060.000000,2664.000000,2837.000000
mean,2009.941176,205.782316,477.930913
std,5.786024,2695.525264,6474.935347
min,2001.000000,0.000000,1.000000
25%,2005.000000,15.295000,34.000000
50%,2009.000000,28.000000,70.000000
75%,2015.000000,41.227500,110.000000
max,2019.000000,64536.000000,121056.000000


**Reasoning**:
Based on the statistical summary, I will identify the minimum, maximum, and average values for 'indice' and 'ranking'. Then, I will find the countries that correspond to the global minimum and maximum values of 'indice' and 'ranking' in the DataFrame.



In [11]:
print("\nAnálisis de la columna 'indice':")
print(f"Valor mínimo de indice: {df['indice'].min()}")
print(f"Valor máximo de indice: {df['indice'].max()}")
print(f"Valor promedio de indice: {df['indice'].mean()}")

print("\nAnálisis de la columna 'ranking':")
print(f"Valor mínimo de ranking: {df['ranking'].min()}")
print(f"Valor máximo de ranking: {df['ranking'].max()}")
print(f"Valor promedio de ranking: {df['ranking'].mean()}")

print("\nPaíses con valores extremos en 'indice':")
min_indice_country = df.loc[df['indice'].idxmin(), 'pais']
max_indice_country = df.loc[df['indice'].idxmax(), 'pais']
print(f"País con el indice más bajo: {min_indice_country} (indice = {df['indice'].min()})")
print(f"País con el indice más alto: {max_indice_country} (indice = {df['indice'].max()})")

print("\nPaíses con valores extremos en 'ranking':")
min_ranking_country = df.loc[df['ranking'].idxmin(), 'pais']
max_ranking_country = df.loc[df['ranking'].idxmax(), 'pais']
print(f"País con el ranking más bajo: {min_ranking_country} (ranking = {df['ranking'].min()})")
print(f"País con el ranking más alto: {max_ranking_country} (ranking = {df['ranking'].max()})")


Análisis de la columna 'indice':
Valor mínimo de indice: 0.0
Valor máximo de indice: 64536.0
Valor promedio de indice: 205.7823160660661

Análisis de la columna 'ranking':
Valor mínimo de ranking: 1.0
Valor máximo de ranking: 121056.0
Valor promedio de ranking: 477.93091293620023

Países con valores extremos en 'indice':
País con el indice más bajo: Dinamarca (indice = 0.0)
País con el indice más alto: Kosovo (indice = 64536.0)

Países con valores extremos en 'ranking':
País con el ranking más bajo: Finlandia (ranking = 1.0)
País con el ranking más alto: Kosovo (ranking = 121056.0)


## Análisis de Datos Faltantes

### Subtask:
Calcular el número total y la proporción de valores nulos para cada columna en el DataFrame `df`, e identificar si alguna columna tiene más del 30% de datos faltantes.


**Reasoning**:
I will calculate the number of null values and their proportion for each column in the DataFrame `df`. Then, I will display these results and identify if any column has more than 30% missing data.



In [12]:
nulos_por_columna = df.isnull().sum()
proporcion_nulos = (df.isnull().sum() / len(df)) * 100

print("Valores nulos por columna:")
print(nulos_por_columna)

print("\nProporción de valores nulos por columna (%):")
print(proporcion_nulos)

columnas_con_muchos_nulos = proporcion_nulos[proporcion_nulos > 30]
if not columnas_con_muchos_nulos.empty:
    print("\nColumnas con más del 30% de datos faltantes:")
    print(columnas_con_muchos_nulos)
else:
    print("\nNo hay columnas con más del 30% de datos faltantes.")

Valores nulos por columna:
codigo_iso      0
anio            0
indice        396
ranking       223
pais            0
dtype: int64

Proporción de valores nulos por columna (%):
codigo_iso     0.000000
anio           0.000000
indice        12.941176
ranking        7.287582
pais           0.000000
dtype: float64

No hay columnas con más del 30% de datos faltantes.


## Verificar Unicidad y Duplicados

### Subtask:
Determinar el número de países y años distintos en el DataFrame, y verificar la existencia de filas completamente duplicadas.


**Reasoning**:
I will determine the number of unique countries and years in the DataFrame `df` as per the first two instructions of the subtask.



In [13]:
print(f"Número de países distintos: {df['pais'].nunique()}")
print(f"Número de años distintos: {df['anio'].nunique()}")

Número de países distintos: 179
Número de años distintos: 17


**Reasoning**:
I need to check for duplicate rows in the DataFrame `df` to complete the subtask. I will use the `.duplicated()` method to identify duplicates and then sum them to get the total count.



In [14]:
num_duplicated_rows = df.duplicated().sum()
print(f"Número de filas duplicadas (exactamente iguales): {num_duplicated_rows}")

Número de filas duplicadas (exactamente iguales): 0


## Validar Consistencia País-Código ISO

### Subtask:
Verificar si existen inconsistencias entre la columna `pais` y su `codigo_iso`, como un mismo código ISO asociado a más de un país.


**Reasoning**:
I will group the DataFrame `df` by `codigo_iso` and count the number of unique countries (`pais`) for each code. Then, I will filter to find `codigo_iso` values that are associated with more than one unique country, storing these inconsistent codes.



In [15]:
inconsistent_codes = df.groupby('codigo_iso')['pais'].nunique()
inconsistent_codes = inconsistent_codes[inconsistent_codes > 1]

print("Códigos ISO con inconsistencias (asociados a más de un país):")
print(inconsistent_codes)


Códigos ISO con inconsistencias (asociados a más de un país):
Series([], Name: pais, dtype: int64)


## Comparación Regional (Bucle For)

### Subtask:
Para cada año, en el subconjunto de países latinoamericanos, identificar el país con el menor y mayor valor de 'indice'.


**Reasoning**:
I need to filter the DataFrame `df` to include only the countries listed in the `america` variable. I will use the `isin()` method on the `codigo_iso` column for this purpose and store the result in `df_america`.



In [16]:
america = ['ARG', 'ATG', 'BLZ', 'BOL', 'BRA', 'CAN', 'CHL', 'COL', 'CRI',
           'CUB', 'DOM', 'ECU', 'GRD', 'GTM', 'GUY', 'HND', 'HTI', 'JAM',
           'MEX', 'NIC', 'PAN', 'PER', 'PRY', 'SLV', 'SUR', 'TTO', 'URY',
           'USA', 'VEN']

df_america = df[df['codigo_iso'].isin(america)].copy()

print("DataFrame filtrado para países latinoamericanos (df_america) creado.")
print(f"Shape de df_america: {df_america.shape}")
df_america.head()

DataFrame filtrado para países latinoamericanos (df_america) creado.
Shape de df_america: (493, 5)


,codigo_iso,anio,indice,ranking,pais
5,ARG,2001,12.0,8.0,Argentina
7,ATG,2001,NaN,NaN,Antigua y Barbuda
20,BLZ,2001,NaN,NaN,Belize
21,BOL,2001,14.5,13.0,Bolivia
22,BRA,2001,18.8,18.0,Brasil


**Reasoning**:
I will iterate through each unique year in `df_america`, filter the data for that year, and then find the countries with the minimum and maximum 'indice' values.



In [17]:
años_unicos = df_america['anio'].sort_values().unique()

print("Comparación Regional (por año) usando un bucle for:")
for anio in años_unicos:
    df_anio_america = df_america[df_america['anio'] == anio].dropna(subset=['indice'])

    if not df_anio_america.empty:
        # País con el menor índice (mayor libertad de prensa)
        min_indice_pais = df_anio_america.loc[df_anio_america['indice'].idxmin()]

        # País con el mayor índice (menor libertad de prensa)
        max_indice_pais = df_anio_america.loc[df_anio_america['indice'].idxmax()]

        print(f"\nAño {anio}:")
        print(f"  Mayor libertad de prensa: {min_indice_pais['pais']} (indice: {min_indice_pais['indice']})")
        print(f"  Menor libertad de prensa: {max_indice_pais['pais']} (indice: {max_indice_pais['indice']})")
    else:
        print(f"\nAño {anio}: No hay datos de 'indice' disponibles para este año en la región.")

Comparación Regional (por año) usando un bucle for:

Año 2001:
  Mayor libertad de prensa: Canadá (indice: 0.8)
  Menor libertad de prensa: Cuba (indice: 90.3)

Año 2002:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 1.0)
  Menor libertad de prensa: Cuba (indice: 97.83)

Año 2003:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 2.0)
  Menor libertad de prensa: Argentina (indice: 35826.0)

Año 2004:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 2.0)
  Menor libertad de prensa: Cuba (indice: 87.0)

Año 2005:
  Mayor libertad de prensa: Bolivia (indice: 4.5)
  Menor libertad de prensa: Cuba (indice: 95.0)

Año 2006:
  Mayor libertad de prensa: Canadá (indice: 4.88)
  Menor libertad de prensa: Cuba (indice: 96.17)

Año 2007:
  Mayor libertad de prensa: Canadá (indice: 3.33)
  Menor libertad de prensa: Cuba (indice: 88.33)

Año 2008:
  Mayor libertad de prensa: Canadá (indice: 3.7)
  Menor libertad de prensa: Cuba (indice: 94.0)

Año 2009:
  Mayor libertad de prensa

## Comparación Regional (Vectorizado)

### Subtask:
Resolver la misma tarea de comparación regional para países latinoamericanos utilizando un enfoque vectorizado con `groupby`, sin usar ciclos explícitos.


**Reasoning**:
I will group `df_america` by 'anio' and then use `idxmin()` and `idxmax()` on the 'indice' column to find the rows corresponding to the countries with the minimum and maximum indices for each year. This will allow for a vectorized approach as requested.



In [18]:
print("Comparación Regional (por año) usando un enfoque vectorizado:")

# Filtrar NaN en 'indice' antes de agrupar para evitar errores en idxmin/idxmax
df_america_filtered = df_america.dropna(subset=['indice'])

# Encontrar el país con el menor índice por año (mayor libertad de prensa)
idx_min = df_america_filtered.groupby('anio')['indice'].idxmin()
paises_mayor_libertad = df_america_filtered.loc[idx_min]

# Encontrar el país con el mayor índice por año (menor libertad de prensa)
idx_max = df_america_filtered.groupby('anio')['indice'].idxmax()
paises_menor_libertad = df_america_filtered.loc[idx_max]

# Mostrar los resultados
for anio in paises_mayor_libertad['anio'].unique():
    mayor_libertad_pais = paises_mayor_libertad[paises_mayor_libertad['anio'] == anio].iloc[0]
    menor_libertad_pais = paises_menor_libertad[paises_menor_libertad['anio'] == anio].iloc[0]

    print(f"\nAño {anio}:")
    print(f"  Mayor libertad de prensa: {mayor_libertad_pais['pais']} (indice: {mayor_libertad_pais['indice']})")
    print(f"  Menor libertad de prensa: {menor_libertad_pais['pais']} (indice: {menor_libertad_pais['indice']})")

# Handle years with no 'indice' data, similar to the loop version
años_con_datos = df_america_filtered['anio'].unique()
años_unicos_total = df_america['anio'].sort_values().unique()

for anio in años_unicos_total:
    if anio not in años_con_datos:
        print(f"\nAño {anio}: No hay datos de 'indice' disponibles para este año en la región.")

Comparación Regional (por año) usando un enfoque vectorizado:

Año 2001:
  Mayor libertad de prensa: Canadá (indice: 0.8)
  Menor libertad de prensa: Cuba (indice: 90.3)

Año 2002:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 1.0)
  Menor libertad de prensa: Cuba (indice: 97.83)

Año 2003:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 2.0)
  Menor libertad de prensa: Argentina (indice: 35826.0)

Año 2004:
  Mayor libertad de prensa: Trinidad y Tobago (indice: 2.0)
  Menor libertad de prensa: Cuba (indice: 87.0)

Año 2005:
  Mayor libertad de prensa: Bolivia (indice: 4.5)
  Menor libertad de prensa: Cuba (indice: 95.0)

Año 2006:
  Mayor libertad de prensa: Canadá (indice: 4.88)
  Menor libertad de prensa: Cuba (indice: 96.17)

Año 2007:
  Mayor libertad de prensa: Canadá (indice: 3.33)
  Menor libertad de prensa: Cuba (indice: 88.33)

Año 2008:
  Mayor libertad de prensa: Canadá (indice: 3.7)
  Menor libertad de prensa: Cuba (indice: 94.0)

Año 2009:
  Mayor libertad

## Tabla Dinámica de Índice Anual por País

### Subtask:
Construir una tabla dinámica para analizar la evolución del índice de libertad de prensa por país a lo largo del tiempo, mostrando el índice máximo anual y reemplazando los valores nulos con cero.


**Reasoning**:
I will create a pivot table as described in the instructions, using 'pais' for rows, 'anio' for columns, 'indice' as values with 'max' aggregation, and filling NaN values with 0.



In [19]:
pivot_df = pd.pivot_table(df, values='indice', index='pais', columns='anio', aggfunc='max', fill_value=0)
print("Tabla dinámica (pivot_df) creada exitosamente:")
pivot_df.head()

Tabla dinámica (pivot_df) creada exitosamente:


anio,2001,2002,2003,2004,2005,2006,2007,2008,2009,2012,2013,2014,2015,2017,2018,2019
pais,,,,,,,,,,,,,,,,
Afghanistán,35.5,40.17,28.25,39.17,44.25,56.50,59.25,54.25,51.67,37.36,37.07,37.44,37.75,39.46,37.28,36.55
Albania,0.0,6.50,11.50,14.17,18.00,25.50,16.00,21.75,21.50,30.88,29.92,28.77,29.92,29.92,29.49,29.84
Alemania,1.5,1.33,2.00,4.00,5.50,5.75,4.50,3.50,4.25,10.24,10.23,11.47,14.80,14.97,14.39,14.60
Algeria,31.0,33.00,43.50,40.33,40.00,40.50,31.33,49.56,47.33,36.54,36.26,36.63,41.69,42.83,43.13,45.75
Andorra,0.0,0.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,6.82,6.82,19.87,19.87,21.03,22.21,24.63


## Análisis de Tabla Dinámica (Extremos)

### Subtask:
Identificar el país con el mayor y menor valor de 'indice' en toda la tabla dinámica resultante.


**Reasoning**:
To identify the country with the maximum 'indice' value, I will find the maximum value across the entire `pivot_df` and then use `stack()` and `idxmax()` to get the corresponding country and year. For the minimum 'indice' value (excluding zeros), I will filter out zero values before finding the minimum and then use `stack()` and `idxmin()`.



In [20]:
max_indice_global = pivot_df.max().max()
min_indice_global = pivot_df[pivot_df > 0].min().min() # Excluir ceros

# Encontrar país y año para el índice máximo
max_indice_pais_anio = pivot_df.stack().idxmax()
max_pais = max_indice_pais_anio[0]
max_anio = max_indice_pais_anio[1]

# Encontrar país y año para el índice mínimo (excluyendo ceros)
min_indice_pais_anio = pivot_df[pivot_df > 0].stack().idxmin()
min_pais = min_indice_pais_anio[0]
min_anio = min_indice_pais_anio[1]

print(f"País con el mayor 'indice' en toda la tabla: {max_pais} en el año {max_anio} con un indice de {max_indice_global}")
print(f"País con el menor 'indice' (distinto de cero) en toda la tabla: {min_pais} en el año {min_anio} con un indice de {min_indice_global}")

País con el mayor 'indice' en toda la tabla: Kosovo en el año 2014 con un indice de 64536.0
País con el menor 'indice' (distinto de cero) en toda la tabla: Austria en el año 2009 con un indice de 0.5


## Análisis de Tabla Dinámica (Años)

### Subtask:
Determinar qué años presentan en promedio los valores de `indice` más altos y los más bajos en la tabla dinámica.


**Reasoning**:
I will calculate the average 'indice' for each year across all countries in the `pivot_df` using `.mean(axis=0)`. Then, I will identify the year with the highest average 'indice' and the year with the lowest average 'indice', and print these results.



In [21]:
promedio_indice_por_anio = pivot_df.mean(axis=0)

anio_mas_alto_promedio = promedio_indice_por_anio.idxmax()
valor_mas_alto_promedio = promedio_indice_por_anio.max()

anio_mas_bajo_promedio = promedio_indice_por_anio.idxmin()
valor_mas_bajo_promedio = promedio_indice_por_anio.min()

print(f"Año con el promedio de 'indice' más alto: {anio_mas_alto_promedio} (Promedio: {valor_mas_alto_promedio:.2f})")
print(f"Año con el promedio de 'indice' más bajo: {anio_mas_bajo_promedio} (Promedio: {valor_mas_bajo_promedio:.2f})")

Año con el promedio de 'indice' más alto: 2013 (Promedio: 449.11)
Año con el promedio de 'indice' más bajo: 2001 (Promedio: 20.03)


## Análisis de Tabla Dinámica (Variabilidad)

### Subtask:
Identificar qué país muestra mayor variabilidad (diferencia entre su máximo y mínimo `indice`) a lo largo del tiempo en la tabla dinámica.


**Reasoning**:
I need to calculate the variability for each country by finding the difference between its maximum and minimum 'indice' values across all years, excluding zero values for the minimum. Then, I will identify the country with the highest variability.



In [22]:
max_indice_por_pais = pivot_df.replace(0, np.nan).max(axis=1)
min_indice_por_pais = pivot_df.replace(0, np.nan).min(axis=1)

variabilidad_por_pais = max_indice_por_pais - min_indice_por_pais

pais_mayor_variabilidad = variabilidad_por_pais.idxmax()
valor_mayor_variabilidad = variabilidad_por_pais.max()

print(f"País con mayor variabilidad en 'indice': {pais_mayor_variabilidad} (Variabilidad: {valor_mayor_variabilidad:.2f})")

NameError: name 'np' is not defined

**Reasoning**:
The previous code failed due to a `NameError` because `numpy` was not imported. I need to import the `numpy` library to make `np` available.



In [23]:
import numpy as np

max_indice_por_pais = pivot_df.replace(0, np.nan).max(axis=1)
min_indice_por_pais = pivot_df.replace(0, np.nan).min(axis=1)

variabilidad_por_pais = max_indice_por_pais - min_indice_por_pais

pais_mayor_variabilidad = variabilidad_por_pais.idxmax()
valor_mayor_variabilidad = variabilidad_por_pais.max()

print(f"País con mayor variabilidad en 'indice': {pais_mayor_variabilidad} (Variabilidad: {valor_mayor_variabilidad:.2f})")

País con mayor variabilidad en 'indice': Kosovo (Variabilidad: 46098.00)


## Análisis de Tabla Dinámica (Constancia y Nulos)

### Subtask:
Identificar países con índice constante a lo largo de todos los años registrados y aquellos que no tienen ningún dato (todos los valores en 0) en la tabla dinámica.


**Reasoning**:
I need to identify countries with a constant index. I will calculate the maximum and minimum index values (excluding zeros) for each country in `pivot_df` and then compare them. If they are equal, the index is constant.



In [24]:
max_val_constant = pivot_df.replace(0, np.nan).max(axis=1)
min_val_constant = pivot_df.replace(0, np.nan).min(axis=1)

# Países con índice constante (donde max == min)
paises_indice_constante = variabilidad_por_pais[variabilidad_por_pais == 0].index.tolist()

print("Países con índice constante (sin variabilidad a lo largo de los años con datos):")
if paises_indice_constante:
    for pais in paises_indice_constante:
        print(f"- {pais}")
else:
    print("Ningún país tiene un índice constante.")

Países con índice constante (sin variabilidad a lo largo de los años con datos):
- Antigua y Barbuda
- Granada


**Reasoning**:
To identify countries with no data, I will sum each row in `pivot_df`. If the sum is zero, it means all values for that country are zero, indicating no data.



In [25]:
paises_sin_datos = pivot_df[pivot_df.sum(axis=1) == 0].index.tolist()

print("\nPaíses sin ningún dato (todos los valores en 0):")
if paises_sin_datos:
    for pais in paises_sin_datos:
        print(f"- {pais}")
else:
    print("Todos los países tienen al menos un dato.")


Países sin ningún dato (todos los valores en 0):
Todos los países tienen al menos un dato.


## Consolidar y Normalizar df_anio

### Subtask:
Consolidar la información de 'libertad_prensa_01.csv' y 'libertad_prensa_02.csv' en un DataFrame llamado `df_anio`. Normalizar los nombres de las columnas a minúscula para asegurar consistencia.


## Limpiar df_codigos

### Subtask:
Explorar 'libertad_prensa_codigo.csv', identificar el código ISO ('ZWE') asociado a dos nombres de país, y eliminar el registro inconsistente ('malo'), conservando solo el válido.


**Reasoning**:
To confirm that the inconsistent entry for 'ZWE' has been successfully removed, I will filter the `df_codigos` DataFrame again to show only the rows where `codigo_iso` is 'ZWE'.



## Resumen:

### Preguntas y Respuestas
No se hicieron preguntas específicas en el planteamiento original del problema; la tarea consistía en realizar un análisis.

### Hallazgos Clave del Análisis de Datos
*   **Consolidación y Limpieza de Datos**: Se consolidaron con éxito los datos de `libertad_prensa_01.csv` y `libertad_prensa_02.csv` en `df_anio` con columnas en minúscula. El archivo `libertad_prensa_codigo.csv` fue limpiado, eliminando una entrada inconsistente ('ZWE' asociado a 'malo'), resultando en un `df_codigos` consistente. Finalmente, `df_anio` y `df_codigos` se unieron para formar el `df` principal.

*   **Estructura del DataFrame**: El DataFrame `df` resultante tiene 3060 filas y 5 columnas (`codigo_iso`, `anio`, `indice`, `ranking`, `pais`). Los tipos de datos son apropiados, con `indice` y `ranking` como `float64`, `anio` como `int64`, y `codigo_iso`, `pais` como `object`.

*   **Resumen Estadístico y Valores Extremos**: El `indice` varía de 0.0 a 64536.0 (promedio 205.78), y el `ranking` de 1.0 a 121056.0 (promedio 477.93). Se identificó a **Dinamarca** con el `indice` más bajo (0.0), y **Finlandia** con el `ranking` más bajo (1.0), indicando mayor libertad de prensa. **Kosovo** presentó los valores más altos en `indice` (64536.0) y `ranking` (121056.0), lo que sugiere una menor libertad de prensa o posibles valores atípicos.

*   **Análisis de Datos Faltantes**: Las columnas `indice` y `ranking` mostraron la presencia de valores nulos (12.94% y 7.29% respectivamente), pero ninguna columna superó el 30% de datos faltantes. Las demás columnas no tenían nulos.

*   **Unicidad y Duplicados**: Se encontraron 179 países distintos y 17 años distintos. No se detectaron filas completamente duplicadas en `df`.

*   **Consistencia País-Código ISO**: Se confirmó la ausencia de inconsistencias; ningún `codigo_iso` está asociado a más de un `pais` diferente en el DataFrame final `df`.

*   **Comparación Regional (Países Latinoamericanos)**:
    *   **Mayor Libertad de Prensa**: Países como **Canadá**, **Trinidad y Tobago**, **Jamaica** y **Costa Rica** mostraron consistentemente los `indice` más bajos (mayor libertad de prensa) en diferentes años.
    *   **Menor Libertad de Prensa**: **Cuba** apareció consistentemente con los `indice` más altos (menor libertad de prensa) a lo largo de casi todos los años analizados.
    *   Se notó un valor anómalo para **Argentina** en 2003 (`indice` de 35826.0), lo que podría ser un error de datos. El año 2016 no tuvo datos de `indice` para la región.

*   **Análisis de Tabla Dinámica (`pivot_df`)**:
    *   **Extremos**: En la tabla dinámica, **Kosovo** mantuvo el `indice` más alto (64536.0 en 2014) y **Austria** el `indice` más bajo (0.5 en 2009, excluyendo ceros).
    *   **Años con Promedios Extremos**: El año **2013** tuvo el promedio de `indice` más alto (449.11), indicando una menor libertad de prensa promedio. El **2001** tuvo el promedio más bajo (20.03), sugiriendo mayor libertad de prensa promedio.
    *   **Variabilidad**: **Kosovo** mostró la mayor variabilidad (`indice` de 46098.00) entre su máximo y mínimo `indice`, lo que podría indicar fluctuaciones significativas o valores atípicos.
    *   **Constancia y Nulos**: **Antigua y Barbuda** y **Granada** fueron los únicos países con un `indice` constante (sin variabilidad) a lo largo de los años con datos. No se encontraron países sin ningún dato (todos los valores en 0) en la tabla dinámica.

### Perspectivas o Próximos Pasos
*   Investigar los valores atípicos, especialmente los altos `indice` para **Kosovo** y **Argentina** (en 2003), para determinar si son errores de datos o reflejan situaciones extremas reales.
*   Profundizar en el análisis de tendencias temporales para países con alta variabilidad o para grupos de países, para entender los factores que influyen en los cambios en la libertad de prensa.

In [26]:
df_codigos[df_codigos['codigo_iso'] == 'ZWE']

,codigo_iso,pais
179,ZWE,Zimbabue
